docs/sql/recursive.md

```diff
@@ -0,0 +1,164 @@
+
+- does not require common-table expressions
+
+- allows mutually-recursive view definitions involving any number of
+  views
+
+- allows much richer queries for defining recursive views, including
+  non-monotone queries.  We eventually aim to support arbitrary
+  queries in recursive views.
+
+With these changes our query language becomes Turing-complete.  As a
+consequence, it is possible to write recursive queries whose
+evaluation never terminates.
+
+## Defining mutually-recursive views
+
+### Forward declarations
+
+Our syntax for mutually-recursive queries is inspired from the C
+programming language.  In C one writes recursive functions by usign
+*forward declarations*.  In a C program the forward declarations have
+to occur *before* the actual function definitions.
+
+For example, to write functions `f` and `g` that invoke each other one
+first writes declarations:
+
+```C
+/* Forward declaration of function g */
+int g(int x);
+
+/* Definition of function f */
+int f(int x) {
+  if (n < 0) return 0;
+  else if (n == 0) return 1;
+  else return n - g(n-1); /* f calling g */
+}
+
+/* Definition of function g */
+int g(int x) {
+  if (n < 0) return 0;
+  else if (n == 0) return 0;
+  else return n - f(n-1); /* g calling f */
+}
+```
+
+Similarly to C, we introduce the notion of *forward view declaration*.
+A forward view declaration looks like a table declaration (but without
+constraints on columns).  The declaration specifies the view name and
+the columns and their types, as in the following example:
+
+```sql
+CREATE RECURSIVE VIEW CLOSURE(x int, y int);
+```
+
+Each view that is involved in a recursive computation needs a forward
+declaration.  The forward declaration must occur before the view
+definition in the program.
+
+### Recursive View definitions
+
+Finally, given a view forward declaration, views can be defined using
+the standard SQL syntax using arbirary query expressions.  The forward
+declaration allows the views to write queries that depend on the
+views themselves:
+
+```sql
+CREATE TABLE EDGES(x int, y int);
+
+-- a local view that depends  on CLOSURE and EDGES
+CREATE LOCAL VIEW STEP AS
+SELECT E.x, CLOSURE.y FROM
+E JOIN CLOSURE ON e.y = CLOSURE.x
+
+-- actual definition of view CLOSURE, which depends on STEP and E
+CREATE MATERIALIZED VIEW CLOSURE AS
+(SELECT * FROM EDGES) UNION (SELECT * FROM STEP);
+```
+
+Notice that view `STEP` does not need a forward declaration, for the
+same reason that function `f` in the C example does not need one.
+However, if view `STEP` is not a `LOCAL` view, then it does need a
+forward declaration -- all views that are user-visible that are
+mutually recursive need forward declarations.
+
+The type inferred by the compiler for the view based on the query must
+match exactly the type of the declared view, including the nullability
+of the columns, and the column names.
+
+### Semantics
+
+Currently the compiler automatically makes all mutually-recursive
```

Comment:
Could help to say what it means to make it distinct (can we say smth like "append DISTINCT" to the query) and also explain why.